### ISSUE TITLE
Models sharing a program can pollute each other's state

### ISSUE THREAD

**Issue by tsherif** [uber/luma.gl#1024]
In the following scenario:
```js
    const sphere = new Model(gl, {
      geometry: new SphereGeometry(),
      program: sharedProgram,
      uniforms: {
        uSampler: texture1
      }
    });

    const cube = new Model(gl, {
      geometry: new CubeGeometry(),
      program: sharedProgram,
      uniforms: {
        uSampler: texture2
      }
    });

	sphere.draw();
	cube.draw();
```
both models will use `texture2` to draw. The reason is that `BaseModel.setUniforms` simply passes the uniforms on to `Program.setUniforms`, overriding whatever was there before. 

The solution would be to have `BaseModel` store uniforms that it sets and always reset them for each draw.

**Comment by ibgreen** [comment]
True, though this is a fundamental limitation of webgl rather than our frameworks. 

Is there some particular reason we want to work around this limitation and provide special support for the program sharing?

Assuming it comes down to the tradeoff of relinking the same shaders several times vs setting uniforms every call?

If we wait till we build out support for uniform buffers the cost of resetting uniforms might be a lot smaller?

FWIW I believe the shader cache already supports program reuse (to avoid relinking already linked programs).

**Comment by tsherif** [comment]
This example explicitly emphasizes this usage: https://github.com/uber/luma.gl/blob/master/examples/lessons/13/app.js

If this isn't usage we want to support, we should remove the example and deprecate passing a program as argument for Model construction.

**Comment by ibgreen** [comment]
> we should remove the example 

How about this: we save the example by just making a 2 line change, create an extra identical program instance and avoid sharing? 

Reasons to keep:

* We currently offer luma.gl ports of all the classic "WebGL lessons", it would seem a little strange not to have all of them there.
* The key feature example 13 is demonstrating is fragment lighting, not program sharing.

> deprecate passing a program as argument for Model construction.

Sure, I'd be fine with this, the `Model` API certainly needs to evolve.

Some quick thoughts:

* There are more conceivable uses for the `program` parameter than sharing programs.
* But I don't see use cases for passing in a pre-created program being very important, given the flexibility of the existing `Model` props.
* Offering the `program` parameter does open the user up to the particular "gotcha" if he/she is not aware of the link between programs and uniforms in WebGL (most users probably aren't).
* And if we do want to share programs later (to cut down on multi-second linking times for identical combos of big shaders) we probably want to do a bigger reimplementation anyway...

**Comment by tsherif** [comment]
The example no longer shares programs, so that's fine.

The problem here is similar to the one with buffers/accessors/attributes. There's no clear sense of who owns the uniforms right now. If I call `Model.setUniforms`, I expect that the uniforms are associated with the model and used for any draws with that model. That is not the case now. If the model's program's uniforms are modified elsewhere (even through Luma's API), draws with that model will be corrupted. 

> And if we do want to share programs later (to cut down on multi-second linking times for identical combos of big shaders) we probably want to do a bigger reimplementation anyway

We should definitely be sharing programs. I believe updating programs is one of the more expensive state changes you can make (I'll see if I can find a reference for this), so avoiding redundant `useProgram` calls would be worthwhile.

**Comment by ibgreen** [comment]
>  The problem here is that there's no clear sense of who owns the uniforms right now. I expect that the uniforms are associated with the model and used for any draws with that model. That is not the case now. If the model's program's uniforms are modified elsewhere (even through Luma's API), draws with that model will be corrupted.

I could agree with this (though I think this unclarity only arises when pushing the API).

My sense is that the current classes like `Model` are intended to be fairly lightweight, and to allow use together with pure webgl classes like Framebuffer, Texture and Program, with all their "warts", requiring users to be webgl savvy.

I wonder if truly systematic management of resource sharing, and also working around of "ownership" issues etc would start to take us into "webgl engine" territory (i.e. we provide a "thick layer" that provides classes at higher levels of abstractions and manages webgl resources transparently below).

While we could decide to evolve the current "core" module into such an "engine", as an option, a webgl engine could also be a new separate submodule (i.e. we might want to leave the current `Model` class as it is for light-weight usage).

While I am not opposed, my personal sense has always been to try to postpone the creation of an engine, thinking that we can get a lot of the perf advantages of the webgl engine by a combination of resource management in deck.gl and "local" improvements in the current luma.gl classes.

But maybe the hour of the "luma.gl webgl engine" has struck at last.

**Comment by tsherif** [comment]
I don't think we have to go that far. Just "least surprise". The following
```
model1.setUniforms(...);
model2.setUniforms(...);
```
doesn't seem like pushing the API, and that they would interact with each other is not intuitive. All that needs to be done to fix this is to have a model track its uniforms and make sure to set them on the program before each draw.

**Comment by ibgreen** [comment]
With "pushing the API" I referred to providing the same program to two `Model`s.

Which illustrates that in my view, the `program` property is an "advanced" parameter, for someone who knows what they are doing and wants to override default behavior. Most users of `Model` would never use the `program` property since the Model auto creates a program using the shader module system. 

And even when used, the `program` property can be used without issue to provide a custom program for that model, without any sharing.

Rather than changing the way Model works these are my suggestions in order of preference:

* Remove the use of `Model.program` property in examples. This is just legacy from the very initial port, serves no purpose, and just makes the examples more complicated.
* Document that the `Model.program` property is an advanced prop and that sharing programs has issues.
* Remove the `Model.program` property entirely, if the fact that it can expose underlying WebGL limitations is still an issue. 

Deeper changes, e.g built-in support for program sharing should be proposed in RFC form so we can look at a complete picture.

**Comment by tsherif** [comment]
Fixed by #1199
